File: src/scanthosegates/campaign/econ/abilities/CryosleeperScanner.java
package scanthosegates.campaign.econ.abilities;

import com.fs.starfarer.api.Global;
import com.fs.starfarer.api.campaign.CampaignFleetAPI;
import com.fs.starfarer.api.campaign.SectorEntityToken;
import com.fs.starfarer.api.campaign.comm.IntelInfoPlugin;
import com.fs.starfarer.api.campaign.comm.IntelManagerAPI;
import com.fs.starfarer.api.impl.campaign.abilities.BaseDurationAbility;
import com.fs.starfarer.api.impl.campaign.ids.Tags;
import com.fs.starfarer.api.ui.TooltipMakerAPI;
import com.fs.starfarer.api.util.Misc;
import org.apache.log4j.Level;
import org.apache.log4j.Logger;
import scanthosegates.campaign.intel.CryosleeperIntel;


public class CryosleeperScanner extends BaseDurationAbility {
    public static String CAN_SCAN_CRYOSLEEPERS = "$CryosleeperScannerAllowed";
    private static final Logger log = Global.getLogger(CryosleeperScanner.class);
    static {log.setLevel(Level.ALL);}

    @Override
    protected void activateImpl() {

    }

    @Override
    protected void applyEffect(float amount, float level) {
        if (Global.getSector().getMemoryWithoutUpdate().getBoolean(CAN_SCAN_CRYOSLEEPERS)){
            for (SectorEntityToken cryosleeper : Global.getSector().getCustomEntitiesWithTag(Tags.CRYOSLEEPER)){
                if (tryCreateCryosleeperReportCustom(cryosleeper, log, true, false)
                        && Global.getSector().getMemoryWithoutUpdate().getBoolean(CAN_SCAN_CRYOSLEEPERS)){
                    Global.getSector().getMemoryWithoutUpdate().set(CAN_SCAN_CRYOSLEEPERS, false);
                }
            }
        }
    }

    @Override
    protected void deactivateImpl() {

    }

    @Override
    protected void cleanupImpl() {

    }

    @Override
    public boolean isUsable() {
        return Global.getSector().getMemoryWithoutUpdate().getBoolean(CAN_SCAN_CRYOSLEEPERS);
    }

    @Override
    public boolean hasTooltip(){return true;}

    public void createTooltip(TooltipMakerAPI tooltip, boolean expanded) {
        CampaignFleetAPI fleet = getFleet();
        if (fleet == null) return;
        tooltip.addTitle("Remote Cryosleeper Survey");
        float pad = 10f;
        tooltip.addPara("Remotely surveys all Cryosleepers in the sector and adds them to the intel screen.", pad);

        if (Global.getSector().getMemoryWithoutUpdate().getBoolean(CAN_SCAN_CRYOSLEEPERS)){
            tooltip.addPara("Cryosleeper survey is ready to activate.", Misc.getPositiveHighlightColor(), pad);
        }
        else {
            tooltip.addPara("Cryosleeper survey has already been used.", Misc.getNegativeHighlightColor(), pad);
        }
        addIncompatibleToTooltip(tooltip, expanded);
    }

    public static boolean tryCreateCryosleeperReportCustom(SectorEntityToken cryosleeper, Logger log, boolean showMessage, boolean listener) {
        if ((!cryosleeper.hasTag(Tags.CRYOSLEEPER) || cryosleeper.hasSensorProfile() || cryosleeper.isDiscoverable()) && listener){
            return false;
        }

        IntelManagerAPI intelManager = Global.getSector().getIntelManager();
        for (IntelInfoPlugin intel : intelManager.getIntel(CryosleeperIntel.class)) {
            CryosleeperIntel cs = (CryosleeperIntel) intel;
            if (cs.getEntity() == cryosleeper) {
                return false; // report exists
            }
        }

        CryosleeperIntel report = new CryosleeperIntel(cryosleeper);
        report.setNew(showMessage);
        intelManager.addIntel(report, !showMessage);
        log.info("Created intel report for cryosleeper in " + cryosleeper.getStarSystem());

        return true;
    }
}

File: src/org/aero/scanThoseGates/campaign/abilities/CryosleeperScanner.kt
package org.aero.scanThoseGates.campaign.abilities

import com.fs.starfarer.api.Global
import com.fs.starfarer.api.campaign.SectorEntityToken
import com.fs.starfarer.api.impl.campaign.abilities.BaseDurationAbility
import com.fs.starfarer.api.impl.campaign.ids.Tags
import com.fs.starfarer.api.ui.TooltipMakerAPI
import com.fs.starfarer.api.util.Misc
import org.aero.scanThoseGates.ModPlugin.Settings.ALLOW_CRYOSLEEPER_SCAN
import org.aero.scanThoseGates.campaign.intel.CryosleeperIntel
import org.apache.log4j.Level
import org.apache.log4j.Logger

class CryosleeperScanner : BaseDurationAbility() {

    override fun applyEffect(amount: Float, level: Float) {
        if (Global.getSector().memoryWithoutUpdate.getBoolean(ALLOW_CRYOSLEEPER_SCAN)) {
            for (cryosleeper in Global.getSector().getCustomEntitiesWithTag(Tags.CRYOSLEEPER)) {
                if (tryCreateCryosleeperReportCustom(cryosleeper, log, showMessage = true, listener = false)
                    && Global.getSector().memoryWithoutUpdate.getBoolean(ALLOW_CRYOSLEEPER_SCAN)
                ) {
                    Global.getSector().memoryWithoutUpdate[ALLOW_CRYOSLEEPER_SCAN] = false
                }
            }
        }
    }

    override fun isUsable(): Boolean {
        return Global.getSector().memoryWithoutUpdate.getBoolean(ALLOW_CRYOSLEEPER_SCAN)
    }
    override fun hasTooltip(): Boolean {
        return true
    }

    override fun createTooltip(tooltip: TooltipMakerAPI, expanded: Boolean) {
        val fleet = fleet ?: return
        tooltip.addTitle("Remote Cryosleeper Survey")
        val pad = 10f
        tooltip.addPara("Remotely surveys all Cryosleepers in the sector and adds them to the intel screen.", pad)
        if (Global.getSector().memoryWithoutUpdate.getBoolean(ALLOW_CRYOSLEEPER_SCAN)) {
            tooltip.addPara("Cryosleeper survey is ready to activate.", Misc.getPositiveHighlightColor(), pad)
        } else {
            tooltip.addPara("Cryosleeper survey has already been used.", Misc.getNegativeHighlightColor(), pad)
        }
        addIncompatibleToTooltip(tooltip, expanded)
    }

    override fun activateImpl() {}
    override fun deactivateImpl() {}
    override fun cleanupImpl() {}

    companion object {
        private val log = Global.getLogger(CryosleeperScanner::class.java)
        init { log.level = Level.ALL }

        fun tryCreateCryosleeperReportCustom(cryosleeper: SectorEntityToken, log: Logger, showMessage: Boolean, listener: Boolean): Boolean {
            if ((!cryosleeper.hasTag(Tags.CRYOSLEEPER) || cryosleeper.hasSensorProfile() || cryosleeper.isDiscoverable) && listener) {
                return false
            }
            val intelManager = Global.getSector().intelManager
            for (intel in intelManager.getIntel(CryosleeperIntel::class.java)) {
                val cs = intel as CryosleeperIntel
                if (cs.entity === cryosleeper) {
                    return false // report exists
                }
            }
            val report = CryosleeperIntel(cryosleeper)
            report.isNew = showMessage
            intelManager.addIntel(report, !showMessage)
            log.info("Created intel report for cryosleeper in ${cryosleeper.starSystem}")
            return true
        }
    }
}
